arviz/utils.py

```diff
@@ -50,14 +54,34 @@ def _var_names(var_names, data):
                 )
             )
 
+        excluded_vars = [
+            var[1:] for var in var_names if var.startswith("~") and var not in all_vars
+        ]
         if excluded_vars:
-            var_names = [i for i in all_vars if i not in excluded_vars]
-
-        existent_vars = np.isin(var_names, all_vars)
-        if not np.all(existent_vars):
+            if (filter == "like") or (filter == "regex"):
+                for pattern in excluded_vars:
+                    if pattern not in all_vars:
+                        if filter == "like":
+                            real_vars = [real_var for real_var in all_vars if pattern in real_var]
+                        else:
+                            # i.e filter == "regex"
+                            real_vars = [
+                                real_var for real_var in all_vars if re.search(pattern, real_var)
+                            ]
+                        excluded_vars.extend(real_vars)
+                        excluded_vars.remove(pattern)
```

Comment:
Back to these two lines. I will try to better define my question, which could then be added as a test (I think). Imagine we have `ds.data_vars = ["theta", "theta_t"]` and we call a given function with `az.xyz(..., var_names="~theta", filter="like")`. We then have that `excluded_vars = ["theta"]` and in the pattern loop: `real_vars=["theta", "theta_t"]` which means that right after extending, `excluded_vars = ["theta", "theta", "theta_t"]` and then the `.remove` deletes the first occurrence of theta. Eventually, as `.remove` only deletes the first occurrence it does work, but it feels very convoluted. I would remove `pattern` at the beginning of the iteration and leave the extend at the end, this would make the operation feel more like _substituting the pattern for the list of matches_ and would not require people reading the code to know that `.remove` deletes only the first occurrence to be able to understand what is happening.

Just realized about the `if pattern not in all_vars` which would make all of the above not happen if `pattern == "theta"` right? Sorry I have not been able to try the code and my _brain python_ tends to miss important details.

Reply:
Yes, you're right, in that case your example would not happen because `theta` is in `all_vars`. However, in this case, as the loop is not executed, only `theta` will be excluded and `theta_t` will still be there.
So the solution to you is to remove the `if pattern not in all_vars` condition, right?

On your second point, I'll indeed move the `.remove` -- or just get rid of it entirely (see Ari's comment above)